Neos.Utility.Arrays/Classes/ValueAccessor.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+namespace Neos\Utility;
+
+/*
+ * This file is part of the Neos.Utility.Arrays package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+/**
+ * Type safe accessing of values from nested arrays
+ */
+class ValueAccessor
+{
+    public function __construct(
+        public readonly mixed   $value,
+        public readonly ?string $pathinfo = null
+    ) {
+    }
+
+    private function createTypeError($message): \TypeError
```

Comment:
I'm not sure whether `\TypeError` is appropriate for this. Besides the semantics, `\TypeError` bypasses Flow's error handling, leading to a rather unpleasant error message like this:

![image](https://github.com/neos/flow-development-collection/assets/2522299/52c23bc3-4548-4879-b93e-a2c69019acfb)

As opposed to:
![image](https://github.com/neos/flow-development-collection/assets/2522299/a822a187-df81-4c09-9569-353683672839)

(using the SPL exception [`\UnexpectedValueException`](https://www.php.net/manual/en/class.unexpectedvalueexception.php) which strikes me as the closest semantic equivalent)

Reply:
@grebaldi Not opinionated there at all. Simply used TypeError as this us what php would throw if one would simply use the return types. Did not much thought into it really.

Would you consider `\UnexpectedValueException` a good solution or should this be a Flow Exception?

Reply:
the neos utility package has no dependency on flow as far as i know.

Reply:
So then let us use \UnexpectedValueException